cmd/flat-network-controller/app/dns_deployment.go

```diff
@@ -0,0 +1,379 @@
+							},
+						},
+						Containers: []corev1.Container{
+							{
+								Name:            "coredns",
+								Image:           "coredns/coredns:1.7.0",
+								ImagePullPolicy: "IfNotPresent",
+								Args: []string{
+									"-conf",
+									"/etc/coredns/Corefile",
+								},
+								VolumeMounts: []corev1.VolumeMount{
+									{
+										Name:      "config-volume",
+										ReadOnly:  true,
+										MountPath: "/etc/coredns",
+									},
+								},
+								Ports: []corev1.ContainerPort{
+									{
+										Name:          "dns",
+										ContainerPort: 53,
+										Protocol:      "UDP",
+									},
+									{
+										Name:          "dns-tcp",
+										ContainerPort: 53,
+										Protocol:      "TCP",
+									},
+									{
+										Name:          "metrics",
+										ContainerPort: 9153,
+										Protocol:      "TCP",
+									},
+								},
+								SecurityContext: &corev1.SecurityContext{
+									Capabilities: &corev1.Capabilities{
+										Add: []corev1.Capability{
+											"NET_BIND_SERVICE",
+										},
+									},
+									ReadOnlyRootFilesystem: &readOnlyRootFilesystem,
+								},
+								LivenessProbe: &corev1.Probe{
+									Handler: corev1.Handler{
+										HTTPGet: &corev1.HTTPGetAction{
+											Path:   "/health",
+											Port:   intstr.IntOrString{IntVal: 8080},
+											Scheme: "HTTP",
+										},
+									},
+									InitialDelaySeconds: 60,
+									TimeoutSeconds:      5,
+									SuccessThreshold:    1,
+									FailureThreshold:    5,
+								},
+								ReadinessProbe: &corev1.Probe{
+									Handler: corev1.Handler{
+										HTTPGet: &corev1.HTTPGetAction{
+											Path: "/ready",
+											Port: intstr.IntOrString{
+												IntVal: 8181,
+											},
+											Scheme: "HTTP",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		if _, err := client.AppsV1().DeploymentsWithMultiTenancy(metav1.NamespaceSystem, net.Tenant).Create(deployment); err != nil {
+			if !errors.IsAlreadyExists(err) {
+				return fmt.Errorf("falied to create deployment %s/%s/%s: %v", net.Tenant, metav1.NamespaceSystem, name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func generateCorefile(network, domainName string) (string, error) {
+	const corefileTeml = `
+    .:53 {
+        errors
+        health
+        ready
+        rewrite name kube-dns.kube-system.svc.{{ .DNSName }}. kube-dns-{{ .Network }}.kube-system.svc.{{ .DNSName }}.
+        rewrite name kubernetes.default.svc.{{ .DNSName }}. kubernetes-{{ .Network }}.default.svc.{{ .DNSName }}.
+        kubernetes {{ .DNSName }} in-addr.arpa ip6.arpa {
+          pods insecure
+          fallthrough in-addr.arpa ip6.arpa
+          ttl 30
+        }
+        prometheus :9153
+        forward . /etc/resolv.conf
+        cache 30
+        loop
+        reload
+        loadbalance
+    }
+`
+
+	tmpl := template.Must(template.New("codefile").Parse(corefileTeml))
```

Comment:
"codefile": should it be "corefile"?

Reply:
yes, should name as "corefile". thanks for pointing it out.